File: code/572.go
/*
 * @Author: lmio
 * @Date: 2023-03-28 15:55:30
 * @LastEditTime: 2023-03-28 16:04:13
 * @FilePath: /leetcode/code/572.go
 * @Description:572. 另一棵树的子树
 */
package code

import ."leetcode/code_struct"

func IsSubtree(root *TreeNode, subRoot *TreeNode) bool {
	if root == nil || subRoot== nil {
			return false
		}
	return  containNode(root, subRoot) ||
			IsSubtree(root.Left, subRoot) ||
			IsSubtree(root.Right, subRoot)
}

func containNode(A *TreeNode, B *TreeNode) bool {
	if A == nil && B == nil {
		return true
	}
	if A == nil || B == nil {
		return false
	}
	return A.Val == B.Val && containNode(A.Left, B.Left) && containNode(A.Right, B.Right)
}

File: code_cpp/572.cpp
#include "utils.h"
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    bool isSubtree(TreeNode *root, TreeNode *subRoot) {
        // 都为空
        if (!root && !subRoot) {
            return true;
        }
        // 其中一个为空
        if (!root || !subRoot) {
            return false;
        }
        if (root->val == subRoot->val &&
            isSametree(root->left, subRoot->left) &&
            isSametree(root->right, subRoot->right)) {
            return true;
        }
        return isSubtree(root->left, subRoot) ||
               isSubtree(root->right, subRoot);
    }

    bool isSametree(TreeNode *r1, TreeNode *r2) {
        if (!r1 && !r2) {
            return true;
        }
        if (!r1 || !r2) {
            return false;
        }
        return r1->val == r2->val && isSametree(r1->left, r2->left) &&
               isSametree(r1->right, r2->right);
    }
};